src/main/ruby/truffleruby/core/time.rb

```diff
@@ -228,6 +228,24 @@ def round(places = 0)
     time + (rounded - original)
   end
 
+  def floor(places = 0)
+    original = to_i + subsec.to_r
+    floored = original.floor(places)
+
+    time = Time.allocate
+    time.send(:initialize_copy, self)
```

Comment:
Why not simply `time = dup` here?

Reply:
Ah, because the spec says so, alright:
```ruby
    it "returns an instance of Time, even if #floor is called on a subclass" do
      subclass = Class.new(Time)
      instance = subclass.at(0)
      instance.class.should equal subclass
      instance.floor.should be_an_instance_of(Time)
    end
```

Reply:
This is a common pattern isn't it? Maybe we should have some kind of `Primitive.dup_as_class(x, T)`

Reply:
Calling `initialize_copy` explicitly seems to be only used for Time so far.
Other classes probably build the new instance directly and pass all required state.
The issue with Time is there isn't a nice constructor to use, they are all very complicated.

Maybe we could have a Truffle::TimeOperations helper, but it seems a bit overkill.